har/har.go

```diff
@@ -0,0 +1,520 @@
+	// if the text field is HTTP decoded (decompressed & unchunked), than
+	// trans-coded from its original character set into UTF-8.
+	Encoding string `json:"encoding,omitempty"`
+}
+
+// NewLogger returns a HAR logger using name and version as the Creator.
+func NewLogger(name, version string) *Logger {
+	return &Logger{
+		creator: &Creator{
+			Name:    name,
+			Version: version,
+		},
+		entries: make(map[string]*Entry),
+	}
+}
+
+// ModifyRequest logs requests.
+func (l *Logger) ModifyRequest(req *http.Request) error {
+	ctx := martian.Context(req)
+	id := ctx.ID()
+
+	return l.LogRequest(id, req)
+}
+
+// LogRequest logs the HTTP request with the given ID. The ID should be unique
+// per request/response pair.
+func (l *Logger) LogRequest(id string, req *http.Request) error {
+	hreq := &Request{
+		Method:      req.Method,
+		URL:         req.URL.String(),
+		HTTPVersion: req.Proto,
+		HeadersSize: -1,
+		BodySize:    req.ContentLength,
+		QueryString: []QueryString{},
+	}
+
+	hreq.Headers = headers(req.Header)
+	hreq.Cookies = cookies(req.Cookies())
+
+	for n, vs := range req.URL.Query() {
+		for _, v := range vs {
+			hreq.QueryString = append(hreq.QueryString, QueryString{
+				Name:  n,
+				Value: v,
+			})
+		}
+	}
+
+	pd, err := postData(req)
+	if err != nil {
+		return err
+	}
+	hreq.PostData = pd
+
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	l.entries[id] = &Entry{
+		ID:              id,
+		StartedDateTime: time.Now().UTC(),
+		Request:         hreq,
+		Cache:           &Cache{},
+		Timings:         &Timings{},
+	}
+
+	return nil
+}
+
+// ModifyResponse logs responses.
+func (l *Logger) ModifyResponse(res *http.Response) error {
+	ctx := martian.Context(res.Request)
+	id := ctx.ID()
+
+	return l.LogResponse(id, res)
+}
+
+// LogResponse logs an HTTP response, associating it with the previously-logged
+// HTTP request with the same ID.
+func (l *Logger) LogResponse(id string, res *http.Response) error {
+	hres := &Response{
+		HTTPVersion: res.Proto,
+		Status:      res.StatusCode,
+		StatusText:  http.StatusText(res.StatusCode),
+		HeadersSize: -1,
+		BodySize:    res.ContentLength,
+	}
+
+	if res.StatusCode >= 300 && res.StatusCode < 400 {
+		hres.RedirectURL = res.Header.Get("Location")
+	}
+
+	hres.Headers = headers(res.Header)
+	hres.Cookies = cookies(res.Cookies())
+
+	mv := messageview.New()
+	if err := mv.SnapshotResponse(res); err != nil {
+		return err
+	}
+
+	br, err := mv.BodyReader(messageview.Decode())
+	if err != nil {
+		return err
+	}
+
+	body, err := ioutil.ReadAll(br)
+	if err != nil {
+		return err
+	}
+
+	hres.Content = &Content{
+		Text:     body,
+		Encoding: "base64",
+		Size:     int64(len(body)),
+		MimeType: res.Header.Get("Content-Type"),
+	}
+
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if e, ok := l.entries[id]; ok {
+		e.Response = hres
+		e.Time = time.Since(e.StartedDateTime).Nanoseconds() / 1000000
+	}
+
+	return nil
+}
+
+// Export returns the in-memory log.
+func (l *Logger) Export() (*HAR, error) {
```

Comment:
Just realized that we can't ever return an error here. Didn't see it when porting it over. Will remove error return value.